### PR TITLE
Google OAuth Sign-In

### PR DESCRIPTION
**Pull request by ZhangCreations**
## Notion ticket link
[Enable users to sign up with google oauth](https://www.notion.so/uwblueprintexecs/Enable-users-to-sign-up-with-google-oauth-4a23bd6c18e440ae94bae28ce6897054)

## Implementation description
Going to explain the data flow since I think that's the best way to explain it.
* Make the user sign in on the front end, from this we obtain a tokenId for that specific user.
* Send tokenId to the back end, Make a call to a Google endpoint to verify the the tokenId and get the user's info, access token, and refresh token.
* Check if the user exists in our database by email, also check if they somehow exist on the Firebase but not locally (this is likely to happen if you drop volumes or smthn)
* If the user doesn't exist, automatically make them an account using data from google
* access & refresh tokens are returned to the frontend and life is good

## Steps to test

There are 4 cases to test here.
1. Email associated with an account that was created with email & password, and the user is now logging in with that email through Google
2. Email does not have account locally or on Firebase
3. There isn't an email associated with the account locally but there is one on Firebase
4. There is an account associated with the email on both Firebase and present locally (This should be the case 99% of the time)

If you were paying attention, you might have noticed that I didn't mention to test if the record exists locally but it's not on Firebase. That's because I don't think that it should ever happen but feel free to let me know if that is a small brain idea.

All this testing should be with a **Google account** (we all have bp emails that work)

1. To test my super swaggy 100% guaranteed to work PR you'll need use the Firebase console.
    a. Go to the planetread-secrets channel on slack and get the credentials for the Google account
    b. Login to [Firebase](https://firebase.google.com/) with those credentials
    c. On the left hand side you should see Authentication, go there and you'll see all the users.
    d. For a given user if you press on the kebab menu on the right side, you'll see the option to delete that user
    e. If you're already created an account using your BP email, delete that user for the sake of testing

2. Run the container, go to the web app, and try logging in with Google.
3. Log out, and log in with Google again. The first time tested case 2, the second time tests case 4.
4. Navigate to the db using the instructions for interesting users in the README for this repo and delete the account that was just made. Login again, this tests case 3.
5. Delete that account again, now create a email & password account with the same email, login with Google, this is case 1.
6. Congratulations, you'e officially exhausted all the cases that I can think of.

## What should reviewers focus on?
* The auth service implementation and the Firebase rest client. These are where the bulk of the work is, especially the auth service. Please let me know if there's a better way to do it than I have right now.
* Themselves, take care of yourself!

cdd

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] For backend changes, I have run the appropriate linters:  `docker exec -it <backend-container-id> /bin/bash -c "black . && isort --profile black ."`
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by gaoxk — CHANGES_REQUESTED** [review]
I can't seem to sign up with a new user on this branch or log in with google- can you double check your rebase didn't cause any funky business? Signup still works correctly on main branch. 
I ran the following commands before attempting tests. Anything else I'm missing?
```
docker-compose down -v
docker-compose up --build
```
I'm seeing this in the docker container logs when I attempt test case 4:
```
ERROR in firebase_rest_client: Failed to sign-in via Firebase REST API, status code = 400 error message = Invalid JSON payload received. Unexpected token.
com', 'requestUri': None, 'returnIdpCred
```
When attempting sign up, gql is returning a generic CORS error.
```
Access to fetch at 'http://localhost:5000/graphql' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```
Thank you for thoroughly explaining your thought process and the data flow for the google sign in - helps me understand this much better!

> 3. There isn't an email associated with the account locally but there is one on Firebase

When would this case happen aside from dev stuff?
> If you were paying attention, you might have noticed that I didn't mention to test if the record exists locally but it's not on Firebase. That's because I don't think that it should ever happen but feel free to let me know if that is a small brain idea.

Correct!

**Comment by ZhangCreations** [comment]
@gaoxk Totally forgot that I added this line into my .env file which is why it isn't working.

`FIREBASE_REQUEST_URI = http://localhost`

Sorry for this oversight.

**Comment by gaoxk** [comment]
Will take another look in a sec!

**Comment by gaoxk** [comment]
Still getting sign-up issues 🤔 added the env var to the root + backend `./env` files.

File: frontend/src/components/auth/Login.tsx [hunk 1]
```diff
@@ -1,10 +1,17 @@
...
+  GoogleLoginResponse,
+  GoogleLoginResponseOffline,
+} from "react-google-login";
 import { Redirect } from "react-router-dom";
-import { gql, useMutation } from "@apollo/client";
 import authAPIClient from "../../APIClients/AuthAPIClient";
 import AuthContext, { AuthenticatedUser } from "../../contexts/AuthContext";
 import ResetPassword from "./ResetPassword";
 
+type GoogleSucces = GoogleLoginResponse | GoogleLoginResponseOffline;
```
Comment (gaoxk) [inline frontend/src/components/auth/Login.tsx:13]: Question: Is GoogleResponse an appropriate name? based on l120, I don't get the impression that this necessarily types a success

File: frontend/src/components/auth/Login.tsx [hunk 2]
```diff
@@ -78,6 +113,19 @@ const Login = () => {
...
           <ResetPassword email={email} />
+
+          <GoogleLogin
+            clientId="175399577852-6hll8ih9q1ljij8f50f9pf9t2va6u3a7.apps.googleusercontent.com"
+            buttonText="Google"
+            onSuccess={(response: GoogleSucces): void => {
+              if ("tokenId" in response) {
+                onGoogleLoginSuccess(response.tokenId);
+              } else {
+                console.log(response);
```
Comment (gaoxk) [inline frontend/src/components/auth/Login.tsx:124]: nit: can you make the error console.logs window alerts instead? that way we'll follow existing login error handeling

Comment (gaoxk) [reply frontend/src/components/auth/Login.tsx:124]: Under what conditions will tokenId be missing? GoogleLoginResponseOffline?

File: backend/python/app/graphql/types/user_type.py [hunk 3]
```diff
@@ -28,6 +28,17 @@ class CreateUserDTO(graphene.InputObjectType):
     approved_languages = graphene.JSONString(required=False, default=None)
 
 
+class CreateUserWithGoogleDTO(graphene.InputObjectType):
+    first_name = graphene.String(required=True)
+    last_name = graphene.String(required=True)
+    role = graphene.Argument(RoleEnum, required=True)
+    email = graphene.String(required=True)
+    uid = graphene.String(required=True)
```
Comment (gaoxk) [inline backend/python/app/graphql/types/user_type.py:36]: nit: would `auth_id` be a clearer name?

File: frontend/src/components/auth/Login.tsx (hunk 2 above)
Comment (ZhangCreations) [reply frontend/src/components/auth/Login.tsx:124]: Yup, not sure when that error is returned though, the way it works is weird. I think it currently has some funky behaviour in incognito modes that I can't quite figure out. Forgot to report that :/

**Review by gaoxk — APPROVED** [review]

**Pull request merged**